test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+
+        proc init=(rhs: [?d] eltType) where d.rank == rank {
+            this.init(d.rank,eltType);
+            this.reshapeDomain(d);
+            this.data = rhs;
+        }
+
+        operator :(from: [?d] ?eltType, type toType: Tensor(d.rank,eltType)) {
+            var t: Tensor(d.rank,eltType) = from;
+            return t;
+        }
+        
+        // ------------ Wasnt sure what to do with these   ----------------
+
+        /*operator =(ref lhs: Tensor(?rank,?eltType), in rhs: ?it) where (isRefIterType(it) || (isArray(rhs) && rhs.eltType == eltType)) && rhs.rank == rank {
+            lhs.reshapeDomain(rhs.domain);
+            lhs.data = rhs;
+        }
+        proc init=(in rhs: ?it) where isRefIterType(it) || isArray(rhs) {
+            this.init(rank,eltType);
+            this.reshapeDomain(rhs.domain);
+            this.data = rhs;
+        }
+        operator :(in from: ?it, type toType: Tensor(?rank,?eltType)) where (isRefIterType(it) || (isArray(from) && from.eltType == eltType)) && from.rank == rank {
+            // compilerError("Cannot convert from ",from.type:string," to ",toType:string);
+            var t: Tensor(rank,eltType) = from;
+            return t;
+        }*/
+
+        operator =(ref lhs: Tensor(?rank,?eltType), itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            lhs.reshapeDomain(itr.domain);
+            lhs.data = itr;
+        }
+
+        proc init=(itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            const A = itr;
+            this.init(A);
+        }
+
+        operator :(itr, type toType: Tensor(?rank,?eltType)) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            var t: Tensor(rank,eltType) = itr;
+            return t;
+        }
+
+
+        // Transposes a vector to a row matrix
+        proc transpose() where rank == 1 {
+            const (p,) = shape;
+            var t = new Tensor(eltType,1,p);
+            t.data[0,..] = this.data;
+            return t;
+        }
+
+        // Matrix transpose
+        proc transpose() where rank == 2 {
+            const (m,n) = this.shape;
+            var M = new Tensor(2,eltType);
+            M.reshapeDomain({0..#n,0..#m});
+            forall (i,j) in M.domain with (ref M, ref this) {
+                M.data[i,j] = this.data[j,i];
+            }
+            return M;
+        }
+
+        // Normalizes the tensor to have unit frobenius norm
+        proc normalize() {
+            const norm = sqrt(frobeniusNormPowTwo(this));
+            const data = this.data / norm;
+            return new Tensor(data);
+        }
+
+        // Retruns new tensor with provided domain
+        proc reshape(dom_) {
+            const dom = domainFromShape((...dom_.shape));
+            var t = new Tensor(dom.rank,eltType);
+            t.reshapeDomain(dom);
+            t.data = for (i,a) in zip(t.domain,this.data) do a;
```

Comment:
Should this be:
```chapel
forall (i, a) in zip(t.domain, this.data) do t.data[i] = a;
```

Reply:
When I change this to `forall`, I get a compiler error.

Reply:
Hmm, you should be able to do that assignment in parallel. What's the error?

Reply:
When I change it to a forall loop, this is the compiler error that I get
```
../lib/Tensor.chpl:238: In method 'reshape':
../lib/Tensor.chpl:241: error: rank mismatch in zippered iteration (can't zip a 3D expression with a 2D array)
  ../lib/Tensor.chpl:248: called as (Tensor(2,real(64))).reshape(dom_: domain(3,int(64),one)) from method 'reshape'
  ../lib/MNISTTester.chpl:75: called as (Tensor(2,real(64))).reshape(shape(0): int(64), shape(1): int(64), shape(2): int(64))
note: generic instantiations are underlined in the above callstack
```

Reply:
Okay, I see. Is this only being used when you want to promote a 1D/2D tensor into a 3D tensor that is degenerate in one dimension? if so, I think you could probably do something like this:
```chapel
proc reshape(dom: domain) {
    var t = new Tensor(dom, eltType);
    t.data[.., .., 0] = reshape(this.data, {dom.dim(0), dom.dim(1)});
    return t;
}
```

Where the inner `reshape` proc is defined in the standard library. Docs here: https://chapel-lang.org/docs/language/spec/arrays.html#ChapelArray.reshape

Reply:
This proc was intended to work for any domain as long as it had the same number of entries as the original tensor. I use this to flatten tensors into 1 dimension, and to reconstruct a 1-tensor into 2 and 3 tensors. I suppose I could use a where clause to distinguish two cases, one where the provided domain rank matches the tensor rank, and one where it does not.

Reply:
I just looked at the `reshape` definition in ChapelArray, and it looks like it also just uses the serial `for`

Reply:
Oh that's unfortunate. I'd probably just leave it as is then if it's gonna be more complex to parallelize.

Reply:
This calls for a comment: using `for` here implies that the generated array is going to be local even if `t.domain` was distributed. Could lead to confusion down the road.